materialize-bigquery/schema.go

```diff
@@ -0,0 +1,191 @@
+	}
+
+	for idx, fieldName := range fields {
+		var (
+			err   error
+			field *bigquery.FieldSchema
+		)
+		field, err = fieldSchema(binding.Collection.Projections, fieldName)
+
+		// This is a fatal error. Panicking.
+		if err != nil {
+			panic(err)
+		}
+
+		table.Schema[idx] = field
+	}
+
+	return table, nil
+}
+
+func (t *Table) Validate(existing *pf.MaterializationSpec_Binding) error {
+	if existing == nil {
+		return sqlDriver.ValidateSelectedFields(ConstraintsForNewBinding(&t.spec.Collection, t.resource.Delta), t.spec)
+	} else {
+		constraints, err := ConstraintsForExistingBinding(existing, &t.spec.Collection, t.resource.Delta)
+		if err != nil {
+			return err
+		}
+		return sqlDriver.ValidateSelectedFields(constraints, t.spec)
+	}
+}
+
+// Whether this Table is configured to be only receive
+// delta updates
+func (t *Table) DeltaUpdates() bool {
+	return t.resource.Delta
+}
+
+// The name of the Table as created in BigQuery
+func (t *Table) Name() string {
+	return t.resource.Table
+}
+
+// fieldSchema inspect each projection and builds up a bigquery.FieldSchema that can then be
+// used within a bigquery.Schema. A field in bigquery is analogous to a column, in RDBMS parlance.
+// Although a slice of projection is passed in, fieldSchema only consumes 1 Projection within the slice
+// that matches the fieldName.
+// A projection will hold an inference types slice that is expected to contains only 1 or 2 values, depending
+// on whether the projection is nullable or not. If the field is nullable, the type slice will contain
+// one entry called "null". Otherwise, the slice needs to contain only 1 element, representing the type
+// for this projection
+func fieldSchema(projections []pf.Projection, fieldName string) (*bigquery.FieldSchema, error) {
+	var fieldType bigquery.FieldType
+	var projection *pf.Projection
+	var includesNull bool
+
+	for _, p := range projections {
+		if p.Field == fieldName {
+			projection = &p
+			break
+		}
+	}
+
+	if projection == nil {
+		return nil, fmt.Errorf("Couldn't find a projection for field name: %s", fieldName)
+	}
+
+	for _, p := range projection.Inference.Types {
+		switch p {
+		// The string format can be further expanded so that it can be stored in a more specific format
+		// When a projection is a string and it has the Inference_String set, it's possible for the user
+		// to pass along more information about the string field to let the connector create a more
+		// appropriate field for that string. If an Inference_String{} is present, it doesn't default back
+		// to a normal string as it is prefered to let the user know that the inference could not be matched.
+		case "string":
+			if projection.Inference.String_ != nil {
+				s := projection.Inference.String_
+				switch s.Format {
+				case "date-time":
+					fieldType = bigquery.DateTimeFieldType
+				case "date":
+					fieldType = bigquery.DateFieldType
+				case "geography-wkt":
+					fieldType = bigquery.GeographyFieldType
```

Comment:
I added this field that didn't exist prior. I just saw it in the BigQuery documentation and thought it could be useful to include it. I can definitively remove it if we don't feel like it's worthwhile to keep it.